src/Microsoft.DocAsCode.MarkdigEngine.Extensions/TripleColon/CodeExtension.cs

```diff
@@ -0,0 +1,258 @@
+                        id = value;
+                        break;
+                    case "highlight":
+                        highlight = value;
+                        break;
+                    case "language":
+                        language = value;
+                        break;
+                    case "interactive":
+                        interactive = value;
+                        break;
+                    default:
+                        logError($"Unexpected attribute \"{name}\".");
+                        return false;
+                }
+            }
+
+            if (string.IsNullOrEmpty(source))
+            {
+                logError($"source is a required attribute. Please ensure you have specified a source attribute");
+                return false;
+            }
+            var (code, codePath) = _context.ReadFile(source, InclusionContext.File, null);
+            if(string.IsNullOrEmpty(code))
+            {
+                logError($"The code snippet \"{source}\" could not be found.");
+                return false;
+            }
+            
+            var updatedCode = GetCodeSnippet(range, id, code, logError).TrimEnd('\r').TrimEnd('\n');
+            updatedCodes.Add(updatedCode);
+
+            if (updatedCode == string.Empty)
+            {
+                return false;
+            }
+
+            if(string.IsNullOrEmpty(language))
+            {
+                language = InferLanguageFromFile(source, logError);
+            }
+
+            htmlAttributes = new HtmlAttributes();
+            htmlAttributes.AddProperty("class", $"lang-{language}");
+            if (!string.IsNullOrEmpty(interactive))
+            {
+                htmlAttributes.AddProperty("data-interactive", language);
+                htmlAttributes.AddProperty("data-interactive-mode", interactive);
+            }
+            if (!string.IsNullOrEmpty(highlight)) htmlAttributes.AddProperty("highlight-lines", highlight);
+
+            RenderDelegate = (renderer, obj) =>
+            {
+                renderer.WriteLine("<pre>");
+                renderer.Write("<code").WriteAttributes(obj).Write(">");
+                renderer.WriteLine(updatedCodes.FirstOrDefault());
+                updatedCodes.RemoveAt(0);
+                renderer.WriteLine("</code></pre>");
+
+                return true;
+            };
+
+            return true;
+        }
+
+        private string InferLanguageFromFile(string source, Action<string> logError)
+        {
+            var fileExtension = source.Split('.').ToList().LastOrDefault();
+            if(fileExtension == null)
+            {
+                logError($"Language is not set, and your source has no file type. Cannot infer language.");
+            }
+            var language = HtmlCodeSnippetRenderer.LanguageAlias.Where(oo => oo.Value.Contains(fileExtension) || oo.Value.Contains($".{fileExtension}")).FirstOrDefault();
+            return language.Key;
```

Comment:
`FirstOrDefault` returns `null` and can cause a `NullReferenceException` in line 137 if the source extension does not exist in our supported list. Consider adding a test case to verify that.

Reply:
Done!